File: Python/nyzostrings/nyzostringmicropay.py
"""
Nyzo String for private Seed
"""

from nyzostrings.nyzostring import NyzoString


__version__ = "0.0.1"


class NyzoStringMicropay(NyzoString):
    def __init__(
        self,
        receiver_identifier: bytes,
        sender_data: bytes,
        amount: int,
        timestamp: int,
        previous_hash_height: int,
        previous_block_hash: bytes,
    ) -> None:
        sender_data = sender_data[:32]
        sender_data_len = len(sender_data)
        bytes_content = bytearray(32 + 1 + sender_data_len + 8 + 8 + 8 + 32)
        bytes_content[0:32] = receiver_identifier
        bytes_content[32] = sender_data_len
        bytes_content[33 : 33 + sender_data_len] = sender_data
        bytes_content[
            33 + sender_data_len : 33 + sender_data_len + 8
        ] = amount.to_bytes(8, byteorder="big")
        bytes_content[
            33 + sender_data_len + 8 : 33 + sender_data_len + 8 + 8
        ] = timestamp.to_bytes(8, byteorder="big")
        bytes_content[
            33 + sender_data_len + 8 + 8 : 33 + sender_data_len + 8 + 8 + 8
        ] = previous_hash_height.to_bytes(8, byteorder="big")
        bytes_content[
            33 + sender_data_len + 8 + 8 + 8 : 33 + sender_data_len + 8 + 8 + 8 + 32
        ] = previous_block_hash
        super().__init__("pay_", bytes_content)
        self.receiver_identifier = receiver_identifier
        self.sender_data = sender_data
        self.amount = amount
        self.timestamp = timestamp
        self.previous_hash_height = previous_hash_height
        self.previous_block_hash = previous_block_hash

    @classmethod
    def from_bytes(cls, byte_buffer: bytes) -> "NyzoStringMicropay":
        buffer = memoryview(byte_buffer)
        receiver_buffer = buffer[:32]
        sender_data_length = min(buffer[32] & 0xFF, 32)
        data_buffer = buffer[33 : 33 + sender_data_length]
        amount = int.from_bytes(
            buffer[33 + sender_data_length : 33 + sender_data_length + 8],
            byteorder="big",
            signed=False,
        )
        timestamp = int.from_bytes(
            buffer[33 + sender_data_length + 8 : 33 + sender_data_length + 8 + 8],
            byteorder="big",
            signed=False,
        )
        previous_hash_height = int.from_bytes(
            buffer[
                33 + sender_data_length + 8 + 8 : 33 + sender_data_length + 8 + 8 + 8
            ],
            byteorder="big",
            signed=False,
        )
        previous_block_hash_buffer = buffer[
            33
            + sender_data_length
            + 8
            + 8
            + 8 : 33
            + sender_data_length
            + 8
            + 8
            + 8
            + 32
        ]
        return NyzoStringMicropay(
            receiver_buffer,
            data_buffer,
            amount,
            timestamp,
            previous_hash_height,
            previous_block_hash_buffer,
        )

    @classmethod
    def from_hex(
        cls,
        receiver_identifier_hex: str,
        sender_data_hex: str,
        amount_hex: str,
        timestamp_hex: str,
        previous_hash_height_hex: str,
        previous_block_hash_hex: str,
    ) -> "NyzoStringMicropay":
        filtered_string = receiver_identifier_hex.replace("-", "")[:64]
        receiver = bytes.fromhex(filtered_string)
        filtered_string = sender_data_hex.replace("-", "")[:64]
        data = bytes.fromhex(filtered_string)
        amount = int(amount_hex, 16)
        timestamp = int(timestamp_hex, 16)
        previous_hash_height = int(previous_hash_height_hex, 16)
        filtered_string = previous_block_hash_hex.replace("-", "")[:64]
        previous_block_hash = bytes.fromhex(filtered_string)
        return NyzoStringMicropay(
            receiver, data, amount, timestamp, previous_hash_height, previous_block_hash
        )

File: JavaScript/src/NyzoStringMicropay.js
// https://github.com/n-y-z-o/nyzoVerifier/blob/master/src/main/java/co/nyzo/verifier/nyzoString/NyzoStringPrefilledData.java

// JS does not support native int64
// An alternative could be https://www.npmjs.com/package/int64-buffer
// Maybe smaller
const Int64 = require('node-int64')

const { NyzoString } = require('./NyzoString')

class NyzoStringMicropay extends NyzoString {

  constructor(receiverIdentifier, senderData, amount, timestamp, previousHashHeight, previousBlockHash) {
    senderData = senderData.slice(0, 32)
    let bytes = Buffer.alloc(32 + 1 + senderData.length + 8 + 8 + 8 + 32)
    Buffer.from(receiverIdentifier).copy(bytes, 0)
    bytes[32] = senderData.length
    Buffer.from(senderData).copy(bytes, 33)
    amount.copy(bytes, 33 + senderData.length)
    timestamp.copy(bytes, 33 + senderData.length + 8)
    previousHashHeight.copy(bytes, 33 + senderData.length + 8 + 8)
    Buffer.from(previousBlockHash).copy(bytes, 33 + senderData.length + 8 + 8 + 8)
    super('pay_', bytes)
    this.receiverIdentifier = receiverIdentifier
    this.senderData = senderData
    this.amount = amount
    this.timestamp = timestamp
    this.previousHashHeight = previousHashHeight
    this.previousBlockHash = previousBlockHash
  }

  getReceiverIdentifier() {
      return this.receiverIdentifier
  }

  getSenderData() {
      return this.senderData
  }

  getAmount() {
      return this.amount
  }

  getTimestamp() {
      return this.timestamp
  }

  getPreviousHashHeight() {
      return this.previousHashHeight
  }

  getPreviousBlockHash() {
      return this.previousBlockHash
  }

  static fromByteBuffer(buffer) {
    const receiverBuffer = buffer.slice(0, 32)
    const senderDataLength = Math.min(buffer[32] & 0xff, 32)
    const dataBuffer = buffer.slice(33, 33 + senderDataLength)
    const amount = new Int64(buffer.slice(33 + senderDataLength, 33 + senderDataLength + 8))
    const timestamp = new Int64(buffer.slice(33 + senderDataLength + 8, 33 + senderDataLength + 8 + 8))
    const previousHashHeight = new Int64(buffer.slice(33 + senderDataLength + 8 + 8, 33 + senderDataLength + 8 + 8 + 8))
    const previousBlockHashBuffer = buffer.slice(33 + senderDataLength + 8 + 8 + 8, 33 + senderDataLength + 8 + 8 + 8 + 32)
    return new NyzoStringMicropay(receiverBuffer, dataBuffer, amount, timestamp, previousHashHeight, previousBlockHashBuffer)
  }

  static fromHex(receiverHexString, dataHexString, amountHex, timestampHex, previousHashHeightHex, previousBlockHashBufferHex) {
    const filteredString = receiverHexString.split('-').join('').slice(0, 64)
    const receiverBuffer = Buffer.from(filteredString, 'hex')
    const dataBuffer = Buffer.from(dataHexString, 'hex')
    const amount = new Int64(amountHex)
    const timestamp = new Int64(timestampHex)
    const previousHashHeight = new Int64(previousHashHeightHex)
    const filteredHash = previousBlockHashBufferHex.split('-').join('').slice(0, 64)
    const previousBlockHashBuffer = Buffer.from(filteredHash, 'hex')
    return new NyzoStringMicropay(receiverBuffer, dataBuffer, amount, timestamp, previousHashHeight, previousBlockHashBuffer)
  }

}


module.exports = {
    version: "0.0.1",
    NyzoStringMicropay
}
